### PR TITLE
Fast-path set_HttpVersion

### PR DESCRIPTION
**Pull request by benaadams**
Use Reference.Equals in fast path for setting `HttpVersion`

Issue

httpversion

Slow path in this change is hardly ever taken as GetKnownVersion returns a string which reference equals (is test for it) and is usually 1.1

/cc @halter73 @cesarbs

### CODE REVIEW

**Comment by benaadams** [comment]
Changed to better fast path

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 1]
```diff
@@ -145,23 +146,39 @@ public string HttpVersion
 
                 return string.Empty;
             }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: nit: add blank line

**Comment by cesarbs** [comment]
Any benchmarks?

**Comment by benaadams** [comment]
@cesarbs its a fairly mild, but positive change

Pre

                   Method |          Mean |      StdDev | Scaled |          RPS |
------------------------- |-------------- |------------ |------- |------------- |
           ParsePlaintext |   927.9056 ns |  13.4866 ns |   1.00 | 1,077,695.89 |
  ParsePipelinedPlaintext |   753.9081 ns |   9.5989 ns |   0.81 | 1,326,421.56 |
          ParseLiveAspNet | 5,015.1654 ns |  71.4417 ns |   5.41 |   199,395.22 |
 ParsePipelinedLiveAspNet | 4,684.1782 ns |  30.5706 ns |   5.05 |   213,484.62 |
             ParseUnicode | 8,448.2189 ns | 145.9482 ns |   9.11 |   118,368.15 |
    ParseUnicodePipelined | 8,247.4369 ns |  95.3297 ns |   8.89 |   121,249.79 |
	
Post
	
                   Method |          Mean |      StdDev | Scaled |          RPS |  Change |
------------------------- |-------------- |------------ |------- |------------- |-------- |
           ParsePlaintext |   926.2544 ns |  12.7442 ns |   1.00 | 1,079,616.99 | + 0.17% |
  ParsePipelinedPlaintext |   740.0802 ns |   6.7339 ns |   0.80 | 1,351,204.93 | + 1.86% |
          ParseLiveAspNet | 4,971.0724 ns |  40.0866 ns |   5.37 |   201,163.84 | + 0.88% |
 ParsePipelinedLiveAspNet | 4,618.2437 ns |  30.0627 ns |   4.99 |   216,532.53 | + 1.43% |
             ParseUnicode | 8,444.6168 ns | 109.6224 ns |   9.12 |   118,418.63 | + 0.04% |
    ParseUnicodePipelined | 8,123.0116 ns |  39.0279 ns |   8.77 |   123,107.05 | + 1.53% |

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 2]
```diff
@@ -145,23 +146,41 @@ public string HttpVersion
...
-                }
                 else
                 {
-                    _httpVersion = Http.HttpVersion.Unset;
+                    HttpVersionSetSlow(value);
                 }
             }
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Would it be that expensive to check for HTTP/1.0 in the fast path? Would inclining such a small method be such a bad thing?

I like the ReferenceEquals part of the change given string.Equals isn't inlined, but I'm not so sure the other parts of the change. I'm a little worried we're overfitting. Bug reports show HTTP/1.0 isn't that uncommon 😢 .

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 3]
```diff
@@ -145,23 +146,41 @@ public string HttpVersion
 
                 return string.Empty;
             }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Was this setter not being inlined previously?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Done

**Comment by benaadams** [comment]
Rebased and squashed to kick off test again (package restore failure)

**Comment by halter73** [comment]
:shipit:

**Comment by halter73** [comment]
Rebase and I'll merge this too.

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
Tests passing

**Comment by halter73** [comment]
Merged: https://github.com/aspnet/KestrelHttpServer/commit/2bf5a966cdf214a365b5c97f8d844924074097a8

**Pull request closed without merging**